x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java

```diff
@@ -0,0 +1,144 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.plan.logical.join;
+
+import org.elasticsearch.xpack.esql.expression.NamedExpressions;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamOutput;
+import org.elasticsearch.xpack.ql.expression.Attribute;
+import org.elasticsearch.xpack.ql.expression.Nullability;
+import org.elasticsearch.xpack.ql.expression.ReferenceAttribute;
+import org.elasticsearch.xpack.ql.plan.logical.BinaryPlan;
+import org.elasticsearch.xpack.ql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class Join extends BinaryPlan {
+
+    private final JoinConfig config;
+    private List<Attribute> lazyOutput;
+
+    public Join(Source source, LogicalPlan left, LogicalPlan right, JoinConfig config) {
+        super(source, left, right);
+        this.config = config;
+    }
+
+    public Join(PlanStreamInput in) throws IOException {
+        super(in.readSource(), in.readLogicalPlanNode(), in.readLogicalPlanNode());
+        this.config = new JoinConfig(in);
+    }
+
+    public void writeTo(PlanStreamOutput out) throws IOException {
+        out.writeSource(source());
+        out.writeLogicalPlanNode(left());
+        out.writeLogicalPlanNode(right());
+        config.writeTo(out);
+    }
+
+    public JoinConfig config() {
+        return config;
+    }
+
+    @Override
+    protected NodeInfo<Join> info() {
+        return NodeInfo.create(this, Join::new, left(), right(), config);
+    }
+
+    @Override
+    public Join replaceChildren(List<LogicalPlan> newChildren) {
+        return new Join(source(), newChildren.get(0), newChildren.get(1), config);
+    }
+
+    public Join replaceChildren(LogicalPlan left, LogicalPlan right) {
+        return new Join(source(), left, right, config);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (lazyOutput == null) {
+            lazyOutput = computeOutput();
+        }
+        return lazyOutput;
+    }
+
+    private List<Attribute> computeOutput() {
+        List<Attribute> right = makeReference(right().output());
+        return switch (config.type()) {
+            case LEFT -> // right side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), makeNullable(right));
+            case RIGHT -> // left side becomes nullable
+                NamedExpressions.mergeOutputAttributes(makeNullable(left().output()), right);
+            case FULL -> // both sides become nullable
+                NamedExpressions.mergeOutputAttributes(makeNullable(left().output()), makeNullable(right));
+            default -> // neither side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), right);
```

Comment:
This is mixing how variable shadowing works with the kind of join. Since we won't have right and full joins for now, I don't think this is useful code, as we cannot properly test it, and it's more subtle than it looks like.